descriptors (getset, wrapper, method), __new__, __init__, builtin_function_or_method, hack js inheritance to match py

s-cork opened the pull request:
It aims to: close #1043, close #1090, close #1075, close #942, close #1093, close #1084, close #1094, close #819,  close #1195

---

**The aim of the pr is to:**
1. improve inheritance of `type objects` (force `type` onto their prototypical chains!)
2. implement `getset_descriptor`, 'wrapper_descriptor', 'method_descriptor'
3. implement `mappingproxy` for `type objects`
4. native implementations of `property`, `classmethod`, `staticmethod`
5. create the type `builtin_function_or_method` aka `Sk.builtin.sk_method`
6. new api for building native typeobjects
---

### inheritance (type and object and maybe metaclasses)

since the implementation of #1131 all type objects are instances of `Sk.builtin.type` and this pr uses the benefits that `prototypical` inheritances brings. 

When doing lookups we no longer have to walk the mro for single inherited objects and do a direct access of the property. 

**Paving the way for metaclasses**:
You can now inherit from type. Though meta classes are not supported since `kwargs` are not supported in the compiler for `class B(metaclass=A)`

```python
class A(type): pass # works fine
>>> A.__mro__
(<class '__main__.A'>, <class 'type'>, <class 'object'>)
```

---
### Multiple inheritance
since multiple inheritance breaks protoypical inheritance I set a new flag `sk$prototypical` on all objects. 
- This flag is created in `$mroMerge` function. 
- If this flag returns `false` we take the slow path in various instances (like `typeLookup`) and check up the `MRO`. 
If this flag is `true` we take the fast path and simply do the equivalent of
```javascript
   if (this.prototype.sk$prototypical) {
        return this.prototype[jsName];
    } else {
        ... // check the MRO
    }
```

---
### `__new__` and `__init__`

- by implementing `tp$init` and `tp$new` we get to be stricter about how the constructor can be used internally. 
- All constructors now enforce `new` as a convention
- python will be forced to use the `tp$call` `tp$new` `tp$init` methods in order to create an object.
- some constructors allow python objects for convenience but this is usually a slower path and not equivalent to python in all cases. e.g. `new Sk.builtin.float_(new Sk.builtin.int_(1))` is valid because `nb$float` is defined. But passing a python string to the float constructor won't work.

---

### getset_descriptors

to implement `getset_descriptors` effectively I move all `klass` attributes onto their `prototype`. 
this improves the logic for `klass.tp$getattr`

implementing `getset_descriptors` allows for the removal of `__class__`, `__name__`, `__bases__`, `__mro__`, `__dict__` from the code base, apart from as `getset_descriptors` on either `Sk.builtin.type` or `Sk.builtin.object`

`tp$mro` is now an array. Since this is critical for `typeLookup` it seems sensible.

Similarly all `sk$klass` objects get a `__dict__` `getset_descriptor` which returns the `$d` for the instance of that `klass`... removing the need to hack `__dict__`

---


### mapping proxy

a `mappingproxy` is a non-writeable `dict` like object. By implementing a `mappingproxy` I remove the `$d` from all `type` objects. This removes the requirement to look inside the `$d` for lookups. When the user asks for `klass.__dict__` they hit the `type` `getset_descriptor` for `__dict__` which gathers up the attributes on the `prototype` and returns a `mapping_proxy`

the `mappingproxy` implementation is close to Cpython and can also be used with the `types` module with tests.

---

### Suggested new API for building native types:

```javascript
Sk.builtin.property = Sk.abstr.buildNativeClass("property", {
    constructor: function property(fget, fset, fdel, doc) {...
    },
    slots: {
        tp$new: Sk.generic.new,
        tp$init: function (args, kwargs) {...},
        tp$descr_get: function (obj, type) {...},
        tp$descr_set: function (obj, value) {...},
        tp$doc: "..."
    },
    methods: {
        getter: {
            $meth: function getter (fget) {...},
            $flags: { OneArg: true },
            $textsig: "...",
            $doc: "..."
        },
        ...
    },
    getsets: {
        __doc__: {
            $get: function () { return this.prop$doc },
            $set: function (value) {...},
            $doc: "..."
        },
        ...
    }
});
```

A limitation of this API is that you can only have one `base` which is true of all builtins. 
there are other options that be tagged onto the object literal including:
- `constructor`: the JS implementation - how to create an instance of the object (defaults to `new Function`)
- `slots`: any slot functions (defining any slot function gives the equivalent dunder method)
- `getsets`: `getset_descriptors` e.g. `__dict__`
- `methods`: e.g. `str.count`
- `classmethods`: e.g. `dict.fromkeys`
- `base`: defaults to `Sk.builtin.object`
- `meta`: defaults to `Sk.builtin.type`
- `proto`: anything here will be added onto the `klass.prototype` 
    - useful for private methods that are neither `slots` nor `getsets` nor `methods`
- `flags`: anything here will be added directly onto the `klass`


---

There are a bunch of other changes that are part of this PR
- move `Sk.builtin.Genric*` to `Sk.generic.*` 
  - a place for templates like `Sk.generic.new`... etc
- `wrapper_descriptors` 
   - eliminates the need to define both the slot and the function on builtin type
     - this is done in `Sk.abstr.setUpSlots`
   - allows for full support of dynamically created magic methods
   - stops the requirement to use dunder methods when slot functions should be used
- `builtin_function_or_method` type (`Sk.buitlin.sk_method`) used by:
  - `method_descriptors` mean the call signature can be defined and separates the `function.tp$call` api from `sk_method.tp$call` api
- showcase `math` with the new api - all `Sk.builtin.func` become `Sk.builtin.sk_method` objects with little fuss.
- showcase `collections` with the new api (including an improved `namedtuple` (close #1040) implementation)
- showcase `itertools` with the new api - these are now all type objects rather than hacked generators
- showcase `operator` with the new api and adding missing features.
- skulpt flags: 
    - `sk$acceptable_as_base_class` - not all builtins are subclassable
    - `sk$object` - a way to determine that you have a python object e.g.`Sk.asserts.assert(obj.sk$object)`
    - `sk$prototypical`
    - `sk$builtinBase` - the most builtin base of an object
   - `sk$baseClass` - a flag on type objects that are direct children of `object`
- all private methods now have `$` rather than also ending in an `_`
- generally building on top of the ideas from #1014

s-cork commented:
Ran speed tests on a version this branch with `fields.py` taking the avg of 10 loops. 

this branch
```
avg in repl: 2.40860002041
avg in browser: 2.670786499977112
```
master
```
avg in repl: 2.813100004196167
avg in browser: 3.221490049362183
```

so approx `17%` improvement...

albertjan commented:
Holy moly that's some improvement @s-cork

albertjan commented:
I'll see if I can find some time today to go through this 😄

s-cork commented:
> I'll see if I can find some time today to go through this 😄

I wouldn't look at this branch! I haven't pushed any changes to this branch since I started hacking too deep! And I went quite deep... 

It's far from ready for review. But if you're curious my current progress i can be found here. It's mostly me hacking away with these ideas, but it does present some nice features that I hope will make it into this pr.
https://github.com/s-cork/skulpt/tree/pr/hacking_for_skulpt

The branch mostly works - (I'm down to to 28 failing py unit 2 tests...)

interesting files to look at:
- https://github.com/s-cork/skulpt/blob/pr/hacking_for_skulpt/src/property_class_static.js
  - native property/classmethod/staticmethod with the new api
- https://github.com/s-cork/skulpt/blob/pr/hacking_for_skulpt/src/descr.js
  - slot_wrappers
  - getset_descripts (aka member_descriptors)
  - method_descriptors
- https://github.com/s-cork/skulpt/blob/pr/hacking_for_skulpt/src/builtindict.js 
  - defines all the `builtin functions` using the new api
- https://github.com/s-cork/skulpt/blob/pr/hacking_for_skulpt/src/sk_method.js
  - the new `builtin_function_or_method` type
- https://github.com/s-cork/skulpt/blob/pr/hacking_for_skulpt/src/slotdefs.js
  - where all the `slot_wrapper` definitions are defined
- https://github.com/s-cork/skulpt/blob/pr/hacking_for_skulpt/src/method.js 
  - the py3 version of `method`
- https://github.com/s-cork/skulpt/blob/pr/hacking_for_skulpt/src/mappingproxy.js
  - builtin classes no longer have dicts we use the mapping proxy
- https://github.com/s-cork/skulpt/blob/pr/hacking_for_skulpt/src/type.js
  - implementation of `type` is changed. it now has the `tp$call` method 
  - which all type object inherit meaning that all type objects need a `tp$new`/`tp$init` method
  - I think a lot of the speed improvements come from the changes in constructor implementations 
    - in `int.js` for example - i'm trying to get to a place where you can only call `new Sk.builtin.int_` with a js integer.
    - leaving `int_.prototype.tp$new` to do all the brunt work of conversion
    - making calls to `new Sk.builtin.int_` by the interpreter super fast.

---

Once you start implementing descriptors the game changes... And I've had a bit too much time... 

At this stage I'm most curious about - what you think of the new api...

albertjan commented:
I'm still trying to get my head around all this but I have a few questions:

- Our main API that is used outside of this repository is probably `Sk.builtin.func` this would continue to work right? `Sk.builtin.func(function() { return Sk.ffi.remapToPy("test"); })`
- a module is still just a function that returns a map?
- What will happen with `Sk.builtin.buildClass`?  

These are the things that are most exposed to the outside and would be most painful to break.

s-cork commented:
In my head i would leave the Sk.builtin.func and Sk.misceval.buildClass api
untouched.

In fact Sk.miscval.buildClass was perfectly primed for this since to build
the class it does Sk.misceval.callsimArray(Sk.builtin.type, args)

Which now uses the tp$call method of type rather than the constructor
function.

Similarly the compiler does tp$call ?
And now all builtin types have that method. So no need to use
applyOrSuspend.

The new api doesn’t replace the existing api by any means, it is a
complement - a way to build a native skulpt type with skulpt slots more
concisely.

With Sk.misceval.buildClass you get functions and attributes. With
Sk.abst.buildNativeClass you get slot_wrappers, method_descriptors and
getset_descriptors.


Internally we can use it. And anyone who wants to build a skulpt_slot based
implementation could use it.


My aim with this pr will be to just show the bare bones - which I hope will
be adapting type, adding descriptors, adding tp$new and tp$init methods and
removing all the slot wrappers that we’ve written internally as
Sk.builtin.funcs


A benefit with this pr will hopefully be that:
Instead of defining nb$add as a js func and then  __add__ as an
Sk.builtin.func, just define the former and let Sk.abstr.setUpSlots do all
the work for you.




On Sat, 30 May 2020 at 21:41, Albert-Jan Nijburg <notifications@github.com>
wrote:

> I'm still trying to get my head around all this but I have a few questions:
>
>    - Our main API that is used outside of this repository is probably
>    Sk.builtin.func this would continue to work right? Sk.builtin.func(function()
>    { return Sk.ffi.remapToPy("test"); })
>    - a module is still just a function that returns a map?
>    - What will happen with Sk.builtin.buildClass?
>
> These are the things that are most exposed to the outside and would be
> most painful to break.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/skulpt/skulpt/pull/1092#issuecomment-636332603>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AIY3YQQ227ODYGCLGBEUSUDRUEEIPANCNFSM4M26EJKA>
> .
>

s-cork commented:
In my head i would leave the `Sk.builtin.func` and `Sk.misceval.buildClass` api
untouched.

In fact `Sk.miscval.buildClass` was perfectly primed for this since to build
the class it does `Sk.misceval.callsimArray(Sk.builtin.type, args)`

Which now uses the `tp$call` method of `type` rather than the `constructor`
function.

Similarly the compiler does `tp$call ? `
And now all `builtin types` have that method. So no need to use
`applyOrSuspend`.


The new api doesn’t replace the existing api by any means, it is a
complement - a way to build a native skulpt type with skulpt slots more
concisely.

With `Sk.misceval.buildClass` you get `functions` and `attributes`. With
`Sk.abst.buildNativeClass` you get `slot_wrappers`, `method_descriptors` and
`getset_descriptors`.


Internally we can use it. And anyone who wants to build a `skulpt_slot` based
implementation could use it.

My aim with this pr will be to just show the bare bones - which I hope will
be adapting `type`, adding `descriptors`, adding `tp$new` and `tp$init` methods and
removing all the `slot wrappers` that we’ve written internally as
`Sk.builtin.funcs`

A benefit with this pr will hopefully be that:
Instead of defining `nb$add` as a js func and then `__add__` as an
`Sk.builtin.func`, just define the former and let `Sk.abstr.setUpSlots` do all
the work for you.

s-cork commented:
Update: I did some more hacking today and my times for `fields.py` for 100 Ticks - avg of 10 loops
```
avg in repl: 1.06900002956
avg in browser: 1.246094465255737
around 60% improvement from master
```
I think this is because I was working on our use of number slots. 
removing the need to check `__add__` etc since we're guaranteed to define `nb$add` if `__add__` is defined and vice versa.
I mean I hope that's what it was - `fields.py` does a lot of number calculations...

albertjan commented:
That's massive @s-cork. 🎉 🚀

s-cork commented:
update on this branch - I now have a version that's passing tests so I've pushed to this branch. It's still quite rough around the edges...

meredydd commented:
This is magnificent stuff -- it makes the type system simultaneously cleaner and more correct, and makes using Skulpt from JS a whole lot easier.

Am I right in thinking this is a breaking change? (At very least, it moves the `Generic*` functions.) I suggest we try to get it reviewed and merged basically as soon as @bnmnetp cuts the next release :)

s-cork commented:
> Am I right in thinking this is a breaking change? (At very least, it moves the `Generic*` functions.) 

I have yet to document anything but things that break - but the short answer is - yes - it's breaking. 

Particularly on constructing a builtin e.g. the only valid call to tuple's constructor is:
```javascript
new Sk.buitin.tuple(Array|undefined)
```
```javascript
Sk.builtin.tuple(Array) -> "assertion fail must be called with 'new'"
new Sk.builtin.tuple(pyIterable) -> "assertion fail must be called with an Array"
/*to convert a pyIterable to a tuple you'd need to use*/
Sk.misceval.callsimArray(Sk.builtin.tuple, [pyIterable])
```

Typically a constructor should only ever be called with a JS object (this improves the speed and is useful for the compiler which already calls the constructor with the JS object). The one exception is `str` mostly because we've used this throughout the codebase as a conversion tool. (But that one is fairly tolerable because all py objects have a `tp$str` method inherited from `object` so `new Sk.builtin.str(pyObject) is really a short hand for `pyObject.tp$str()`)

---
Other breaking changes off the top of my head

`Sk.builtin.object.prototype.GenericGetAttr -> Sk.generic.getAttr`
`Sk.builtin.object.prototype.GenericSetAttr -> Sk.generic.setAttr`
`Sk.builtin.object.prototype.GenericPython(Get/Set)Attr -> removed`
`Sk.builtin.type.typeLookup -> removed`
all private methods on builtins (will) now have `$` rather than a trailing underscore e.g.
`Sk.builtin.list.prototype.list_sort_ -> Sk.builtin.list.prototype.$list_sort`

(not especially breaking but notable)

- Py2 methods -> Py3 methods 
  (unbound methods are no longer a thing, so `im_klass` is no longer a thing)
- Old style classes no longer supported - 
  (all classes inherit from object in py2 and py3)

---

I'll try and separate some logic in this pr and create a separate pr for small non breaking changes that can be pre-release changes - e.g. everywhere I found we weren't using `new` but should have been.

s-cork commented:
a very minor detail - related to implementing slots in this pr - in python 3.8:
```
bpo-36793: Removed __str__ implementations from builtin types bool, int, float, complex and a few classes from the standard library. They now inherit __str__() from object.
```
This seems sensible to me. removes some repeat code in the code base. I'll put this into this pr unless any strong objections. Only changes I think 1 test in our library for multiple inheritance where it should override `__repr__` instead of `__str__`

s-cork commented:
An update on this branch. I've recently been working on int and unifying int and long. 

I've got it to a place where py2 compatibility is maintained by making `long` a trivial subclass of int.
There were a few tests where arithmetic operations on longs may become ints in py2 mode. 
In py3 subclassing from int should now work as expected. 

I used `JSBI` over `biginteger`. https://github.com/GoogleChromeLabs/jsbi 

---

**Things I'm struggling with:**
Closure compiler - @rixner  - I saw you did some work on this in your biginteger pr request but I wasn't sure how to achieve what you did. 
I couldn't actually fully remove biginter without closure complaining. 
And similarly it complains for jsbi. 


I'm doing well with speed. `fields.py` is now consistently running at `0.6 seconds` on this branch (in the repl). Which is around `80%` speed improvement on master! and only 3 times slower than Cpython. 

though I have an oddity that if I remove a function on `Sk` like `Sk.setupDictIterators` which is no longer called in the codebase - the speed becomes twice as slow. And I've no idea why. 🤷‍♂️ 

---

lots left to do... but just in case you felt this was going cold.

s-cork commented:
Another update on this branch.
I've now merged master into this branch and am relatively close to a pull request that's worth reviewing. 
I've merged #1013 into this branch. 
As well as fixing #949 using the ideas @acbart  suggested.  

I've got a slightly adapted tact for name mangling. You can see the approach in the string constructor. There also seemed to me, no reason to treat mangling any different for names/words so I unified the approach.

---
**Speed update**
I upped the `TICKS` in the `fields.py` test from `100` to `1000` in the repl taking the average of as many rounds as I was prepared to wait... 
```
this_branch: 4.4  seconds (repl py3)
master:      28.5 seconds (repl py3)
Cpython:     2.2  seconds (ipython terminal macos)
Brython:     130  seconds (console firefox)
```
so overall approx `85%` improvement on master and running at `2x Cpython` 🔥

s-cork commented:
Branch update and questions:
- How can I make this branch easier to review?
- Where should I outline the breaking changes?


---

I've documented the slots using jsdoc syntax as best I could... 
I've tried map the slot to the dunder, provide expected parameters and what the type of the return value should be particularly if it is a JS object e.g.
```
/**
 * @method sq$contains
 * 
 * @param {pyObject} key
 * @param {boolean=} canSuspend
 * 
 * @implements __contains__
 * @returns {boolean}
*/
```
![Screen Shot 2020-06-26 at 12 14 24](https://user-images.githubusercontent.com/36813890/85819582-af328800-b7a6-11ea-8853-73b6906cea72.png)


I've added a section in `hacking.md` about how to build a native skulpt class using the new api. 
(I think much of `hacking.md` could be updated - but for another time...)

s-cork commented:
Branch update and questions:
- How can I make this branch easier to review?
- Where should I outline the breaking changes?


---

I've documented the slots using jsdoc syntax as best I could... 
I've tried map the slot to the dunder, provide expected parameters and what the type of the return value should be particularly if it is a JS object e.g.
```
/**
 * @method sq$contains
 * 
 * @param {pyObject} key
 * @param {boolean=} canSuspend
 * 
 * @implements __contains__
 * @returns {boolean}
*/
```
![Screen Shot 2020-06-26 at 12 14 24](https://user-images.githubusercontent.com/36813890/85819582-af328800-b7a6-11ea-8853-73b6906cea72.png)


I've added a section in `hacking.md` about how to build a native skulpt class using the new api. 
(I think much of `hacking.md` could be updated - but for another time...)

acbart commented:
I don't want to interfere with the review process, but I've been following these changes and I'm extremely excited to test them in my form. Is this PR more or less ready?

albertjan commented:
@acbart feel free to review parts or tell us about what you encounter.

s-cork commented:
@acbart I’d say yes more or less ready. It’d be great to hear how it does in the wild. One of my main aims has been not to break anything...

acbart commented on src/builtin.js:
```diff
@@ -1151,34 +823,27 @@ Sk.builtin.filter = function filter (fun, iterable) {
-        if (obj.tp$getattr(attr)) {
-            return Sk.builtin.bool.true$;
-        } else {
+    const res = obj.tp$getattr(pyName, true);
```
I think this has to be surrounded with a Try/Catch to catch any AttributeError that might occur in a descriptor. This mirrors the behavior of CPython's implementation a little more closely: https://gist.github.com/taozle/f3b3d90337bfde97be68858ea89c66b7

At least, adding the following code seems to fix an issue with one of my weirder classes (which overrides every dunder method in Python).

```javascript
let res;
    try {
        res = obj.tp$getattr(pyName, true);
    } catch (e) {
        if (e instanceof Sk.builtin.AttributeError) {
            return false;
        } else {
            throw e;
        }
    }
```

s-cork replied:
ta - curiosity did you override `__getattribute__`/`__getattr__` as dunder function with `Sk.builtin.func` or did you directly override `tp$getattr` with a js func?

I've added in the try catch and made it suspension aware.

acbart replied:
It's a [Python class](https://github.com/pedal-edu/pedal/blob/master/pedal/sandbox/result.py), not one written in JavaScript.

acbart commented on src/dictviews.js:
```diff
@@ -0,0 +1,198 @@
+        },
+
+        sq$length: function () {
+            this.dict.get$size();
```
Missing a `return` here.

s-cork commented:
A note on closure compiler.

so it turns out the reason I couldn't remove `biginteger.js` before was because 
```javascript
/**
 * @fileoverview
 * @suppress {checkTypes}
 */
```
was included in that file - which essentially meant that closure was not checking types at all. 
Which i guess is what #1044 came across and did a lot of good things to fix. 

I went through this PR and fixed all the `checkType` errors within `Sk.builtin`, `Sk.abstr`, `Sk.misceval` namespaces that either resulted from this PR or had been inherited.

I left the `checkType` errors that were in `ast.js`, `compile.js` and other similar compile files since #1044 has done much of the work to fixe these. 

of which there are the following number of errors:
- 7 - `tonkenize.js`
- 1 - `parse_tables.js`
- 1 - `parser.js`
- 1 - `ast_nodes.js`
- 55 - `ast.js`
- 14 - `compile.js`

There are similar errors for the packages that we rely on - `jsbi`, `strftime`, `setimmediate`, `process`

For now I've moved `checkTypes` from `jscomp_error` to `jscomp_off` in `webpack.config.js`

acbart commented:
I haven't done too much stress testing, but I've run the latest changes through BlockPy (which does a lot of funky stuff) and it seems to be working with the bug fixes I mentioned. I didn't see a massive speedup in the loading of our massive library, which is our major blocking issue - although thanks to one of the smaller changes (Sk.sysmodules being a proper dictionary) I did manage to make some gains there. That aside, this really seems to simplify the codebase. At some point by the end of the summer, hopefully I will have a chance to rewrite some of our libraries to use the new style. That's kind of far away right now, though, so I wouldn't wait on me. But either way, this seems pretty cool!

albertjan commented on src/type.js:
```diff
@@ -1,650 +1,338 @@
-        }
-    }
-};
+if (Sk.builtin === undefined) {
```
it might be good to move this somewhere else, it's a bit odd it lives here.

s-cork replied:
what about adding this to `/utils.js` - i presume it's defined here since this is the first place that the `Sk.builtin` namespace is used...

albertjan commented on src/type.js:
```diff
@@ -710,29 +384,219 @@ Sk.builtin.type.buildMRO_ = function (klass) {
+        bases.push(Sk.builtin.object);
+    }
+    function solid_base(type) {
+        // if we support slots we would need to change this function - for now it just checks for the builtin.
```
Could you explain this? You're pretty far along implementing slots right?

s-cork replied:
Yes. So weirdly the best base is not always the first bases in the tuple of bases! Who knew...

Eg
class A: pass
class B(A, int): pass
B.__base__ = int

This algorithm for finding best base follows a simplified version of the cpython algorithm.

Currently the solid base is just the first builtin we find. The solid base is what is required to check for inheritance conflicts. 

When slots are in the game then the solid base will be something like the first class we find with slots. 

There are certain rules about inheriting from classes with slots that do and don’t cause layout conflicts... this algorithm would deal with those layout conflicts eg. 

class A: __slots__= ‘foo’
class B: __slots__= ‘bar’
class C(A,B): pass # error layout conflict 

For now the only layout conflicts we can have are conflicts from inheriting from multiple builtins. 

class A(dict): pass
class B(int): pass
class C(A,B): pass # layout conflicts the solid bases of A and B are not compatible.

s-cork replied:
Here are the cpython versions 


https://github.com/python/cpython/blob/4a21e57fe55076c77b0ee454e1994ca544d09dc0/Objects/typeobject.c#L2091


https://github.com/python/cpython/blob/4a21e57fe55076c77b0ee454e1994ca544d09dc0/Objects/typeobject.c#L2012

s-cork replied:
ps I think I unwittingly hit resolve on a few of these by mistake on my phone. And hence the unformatted code snippets! Sorry.

albertjan replied:
python is always weirder then you expect, when you look at it as closely as we do when making skulpt. 😄 this is such a good example of that

albertjan commented:
I can't say I've fully reviewed a PR with 17000 changed lines though. In tests we trust though as we did in previous big upgrades.

s-cork commented:
Thanks @albertjan i’ll go through those later.

rixner commented:
I strongly feel that we need to take care of all of the pending pull requests before we even consider this, as it is such an invasive change.  @albertjan I will take a look at it, but I fear that it's going to break my code base and a lot of others...

s-cork commented on src/compile.js:
```diff
@@ -291,7 +194,7 @@ Compiler.prototype.makeConstant = function (rest) {
     this.u.consts[v] = val;
     return v;
-}
+};
```
I think my auto formatter did it...

s-cork replied:
yeah prettier decided that the following should happen... 

```
function foo () {
...
}
var foo = function () {
...
};
```

s-cork commented on src/compile.js:
```diff
@@ -902,16 +795,22 @@ Compiler.prototype.vexpr = function (e, data, augvar, augsubs) {
             switch (e.ctx) {
                 case Sk.astnodes.AugLoad:
-                    out("$ret = Sk.abstr.gattr(", augvar, ",", mname, ", true);");
+                    out("$ret = ", augvar, ".tp$getattr(", mname, ", true);");
```
This was performance reasons. Since every python now inherits from object we know there will be a tp$getattr method. And since it’s such a hot function calling it directly seemed worthwhile.  It’s similar to how we use tp$call directly in compile code if it exists.

s-cork commented on src/enumerate.js:
```diff
@@ -1,66 +0,0 @@
- * @param {number=} start
- * @extends Sk.builtin.object
- */
-Sk.builtin.enumerate = function (iterable, start) {
```
You found it. But yes I put these all in iterator_objects.js. Mostly because when I was working on the interface they were so similar. I can put things back into their own files if preferred.

s-cork commented on src/frozenset.js:
```diff
@@ -1,421 +0,0 @@
-/**
```
I moved this to set.js and did a refactor as frozen set now just uses the set methods. Cpython does the same and not set and frozen set live in setobject.h.

s-cork commented on webpack.config.js:
```diff
@@ -37,13 +37,14 @@ module.exports = (env, argv) => {
             minimizer: [
                 new ClosureWebpackPlugin({mode: 'STANDARD'}, {
-                    jscomp_error: ['accessControls', 'checkRegExp', 'checkTypes', 'checkVars',
+                    jscomp_error: ['accessControls', 'checkRegExp', 'checkVars', /*'checkTypes',*/
```
I left it commented as a reminder that we are aiming for this to be in jscomp_error but it’s currently not feasible...

s-cork commented on src/builtin.js:
```diff
@@ -908,70 +566,73 @@ Sk.builtin.hash = function hash (value) {
     return Sk.misceval.chain(Sk.importModule("sys", false, true), function (sys) {
         if (Sk.inputfunTakesPrompt) {
-            return Sk.misceval.callsimOrSuspendArray(Sk.builtin.file.$readline, [sys["$d"]["stdin"], null, lprompt]);
+            return Sk.builtin.file.$readline(sys["$d"]["stdin"], null, lprompt);
```
Chain handles the suspension already. Here we directly call the function since we can. $readline is a js function and so no reason to use callsim here.

s-cork commented on src/builtin.js:
```diff
@@ -1492,23 +1025,23 @@ Sk.builtin.next_ = function next_ (iter, default_) {
+Sk.builtin.vars = function vars() {
     throw new Sk.builtin.NotImplementedError("vars is not yet implemented");
 };
-Sk.builtin.xrange = Sk.builtin.range;
```
Yes that’s handled as before in builtindict.js

s-cork commented on src/compile.js:
```diff
@@ -126,115 +126,18 @@ Compiler.prototype.niceName = function (roughName) {
-
 function unfixReserved(name) {
-    return name.replace(/_\$r[wn]\$$/, "");
+    return name.replace(/_\$rw\$$/, "");
```
Yeah this is the change in #1117

s-cork commented on webpack.config.js:
```diff
@@ -37,13 +37,14 @@ module.exports = (env, argv) => {
                                    'nonStandardJsDocs', 'strictModuleDepCheck', 'undefinedVars',
                                    'unknownDefines', 'visibility'],
-                    jscomp_off: ['fileoverviewTags', 'deprecated'],
+                    jscomp_off: ['fileoverviewTags', 'deprecated', 'uselessCode', 'suspiciousCode', 'checkTypes',],
```
Yeah. See my comment about type checks. There are a lot of type errors on master that weren’t feasible to fix in this pr. 

Type errors:
7 - tonkenize.js
1 - parse_tables.js
1 - parser.js
1 - ast_nodes.js
55 - ast.js
14 - compile.js

I’m doing some learning on web pack because I currently have no real clue about it. At present jsbi and other libraries we use don’t pass type checking either. 

The reason jsbi wasn’t passing was because the node_module version is already minified- when I recently copied over the full version from their GitHub and let our compiler do the minifying and that seemed to work.

s-cork commented on support/externs/sk.js:
```diff
@@ -1,2 +1,49 @@
 // Tell closure compiler this is defined!
 var Sk = {};
+
```
Yes - this was temprorary while i was trying to fix type checking. I needed to extern jsbi while fixing other errors. I’ll sort this one out when I can and report back.

s-cork commented on src/function.js:
```diff
@@ -233,92 +25,153 @@ Sk.exportSymbol("Sk.builtin.checkFunction", Sk.builtin.checkFunction);
+        __defaults__: {
+            $get: function () {
+                return new Sk.builtin.tuple(this.$defaults);
+            }, // technically this is a writable property but we'll leave it as read-only for now
```
I know right! I only found it out when hacking around with this pr.

s-cork replied:
it's technically correct to have a `__closure__` getset descriptor...

```python
# from pypy which also uses getset_descriptor
mappingproxy({'__new__': <staticmethod at 0x10a565bb0>,
              '__call__': <function builtin_function.__call__(self, *args, **keywords)>,
              '__get__': <function function.__get__(function, obj, cls=<no value>)>,
              '__repr__': <function function.__repr__(self)>,
              '__code__': <getset_descriptor at 0x10a5ce408>,
              '__doc__': <getset_descriptor at 0x10a53ca88>,
              '__name__': <getset_descriptor at 0x10a5ce460>,
              '__qualname__': <getset_descriptor at 0x10a5ce4b8>,
              '__dict__': <getset_descriptor at 0x10a5ce510>,
              '__defaults__': <getset_descriptor at 0x10a5ce590>,
              '__defaults_count__': <getset_descriptor at 0x10a5ce618>,
              '__kwdefaults__': <getset_descriptor at 0x10a5ce698>,
              '__annotations__': <getset_descriptor at 0x10a5ce6f0>,
              '__globals__': <getset_descriptor at 0x10a5ce770>,
              '__closure__': <getset_descriptor at 0x10a5ce7f0>,
              '__module__': <getset_descriptor at 0x10a5ce848>,
              '__weakref__': <getset_descriptor at 0x10a5e79b0>})
```

s-cork commented on src/iteratorobjects.js:
```diff
@@ -0,0 +1,290 @@
+ * @param {number|string=} start
+ * @extends Sk.builtin.object
+ */
+Sk.builtin.enumerate = Sk.abstr.buildIteratorClass("enumerate", {
```
Should it live here?

s-cork commented on src/lib/copy.py:
```diff
@@ -17,7 +17,7 @@ def copy(x):
     if copier:
         return copier(x)
-    if cls in (type(None), int, float, bool, long, str, tuple, type):
+    if cls in (type(None), int, float, bool, str, tuple, type):
```
Great question. I couldn’t think of a way to simultaneously remove long from python3 mode and keep this piece of code??

s-cork replied:
Idea. What if I do a jseval inside the scope of this function. 

Something akin to
long = jseval(“Sk.builtins.long || Sk.builtins.int”)

s-cork replied:
Another idea. I can add a check for isinstance since now long is an instance of int then it will be covered by that case... or __name__ == “long” some. Maybe that’s better? 🤔

s-cork commented on src/module.js:
```diff
@@ -1,19 +1,51 @@
-Sk.builtin.module.prototype.tp$getattr = Sk.builtin.object.prototype.GenericGetAttr;
-Sk.builtin.module.prototype.tp$setattr = Sk.builtin.object.prototype.GenericSetAttr;
-Sk.builtin.module.prototype.tp$name = "module";
+Sk.builtin.module = Sk.abstr.buildNativeClass("module", {
```
The api?

albertjan replied:
The fact that a module is a real thing now not a completely empty class.

s-cork replied:
Ah yes and it leads nicely to dealing with $d in the constructor as opposed to in the import maybe. 

I think there’s another pr I remember seeing somewhere that looked at this. I think it was in #991

Surprisingly module is technically subclassable. Though I don’t know how that would be done since it’s not in python’s `__builtins__`

acbart replied:
I think this breaks down when you `setattr` a `__getattr__` onto a newly created module object. My library does this for mocking purposes, creating fake versions of other modules with special `__getattr__` functions that return data. I'm not clear what the proper solution is - it seems like the `tp$getattr` would need to check to see if `$d` has a `__getattr__` field, and then call that function instead - but my own tests with that don't work for cases where the returned field is a function. Any insight about this?

acbart replied:
Okay, turns out some other code had a bug; the code below actually seems to work. However, I'm pretty sure it's not at all a good way to handle this. How should this correctly check the existence of a custom __getattr__ that is given through `setattr`?

```javascript
let customGetAttr = this.$d["__getattr__"];
if (customGetAttr) {
    const ret = Sk.misceval.callsimArray(customGetAttr, [pyName]);
    if (ret !== undefined) {
        return ret;
    }
}
```

s-cork replied:
here's the pep for this right? https://www.python.org/dev/peps/pep-0562/

I think you're code looks good to me... so long as it happens after failing to find an `object` in the `module` namespace in line with the pep... It probably needs some suspension awareness thrown in since it seems the perfect place to lazy load something and make a call to the server... which is one of the intended use cases from the pep...

in this pr...
`setattr` will call `Sk.gener.setAttr/object.prototype.tp$setattr` 
so in this case `__getattr__` should be added to `$d` on the module object... and then later called by `module.prototype.tp$getattr` in the expected way... I expect?

_(nb: I have intend to customise the `tp$setattr` for `modules` since in this pr it becomes the only `skulpt object` where `$d` is a javascript object and not a `skulpt dict` so seems sensible to have a custom version)_

p.s. if you find bugs shout so I can fix them!
p.p.s. please don't rely on this pr! My current approach is to make incremental changes that are non breaking - until they're exhausted (or I am 😉 )... I doubt this pr will ever get merged in it's current form...

s-cork commented on src/builtin.js:
```diff
@@ -360,41 +228,39 @@ Sk.builtin.min = function min($default, key, args) {
         }
     } else {
-        let lowest_compare = Sk.misceval.callsimOrSuspendArray(key, [lowest]);
+        let lowest_compare = Sk.misceval.callsimArray(key, [lowest]);
```
No. I wrote the original pr without knowing what suspensions were!

My understanding now is that misceval.callsimOrSuspendArray just allows a a suspension to be returned. But it doesn’t handle the suspension. 

Misceval.chain is used to actually handle a retuned suspension. And since we don’t use chain here we can’t handle a suspension. 

I can rewrite this algorithm as separate pr to handle suspensions? now that I know what they are...!

s-cork replied:
I have a pr for this see #1119 
I will merge those changes into this branch when/if they are merged into master.

s-cork commented on src/builtin.js:
```diff
@@ -360,41 +228,39 @@ Sk.builtin.min = function min($default, key, args) {
+    if (!nargs) {
+        throw new Sk.builtin.TypeError("max expected 1 argument, got 0");
     }
+    const default_key = Sk.abstr.copyKeywordsToNamedArgs("min", ["default", "key"], [], kwargs, [null, Sk.builtin.none.none$]);
```
I put comments above the original function which should make it into jsdocs and to the website.

s-cork replied:
No but I did add jsdoc comments to that function with an example so it should get compiled to the website docs

s-cork commented on src/builtin.js:
```diff
@@ -416,111 +282,68 @@ Sk.builtin.max = function max($default, key, args) {
+            tot = new Sk.builtin.float_(Sk.builtin.asnum$(tot));
         }
-
+        // else if (i instanceof Sk.builtin.lng) {
```
Yes in py2 mode. This algorithm doesn’t need to check it though since it just uses the nb$add architecture. 

Also long in py2 mode is just a trivial subclass of int_ 
You can now find long  defines at the end of of int.js

s-cork replied:
actually technically this function is already broken.
I'll submit a different pr for this since it's just dealing with floats as being the problem. 
It fails at complex numbers already. 
On master branch:
```python
>>> sum([complex(1,2), 1])
# TypeError: unsupported operand type(s) for +: 'int' and 'complex'
```

s-cork replied:
resolved
see commit 7bbd4b4ef2826831bdc4526174ffe129b80b54c0
and pr #1129

s-cork commented on src/ast.js:
```diff
@@ -2716,7 +2716,7 @@ function parsenumber (c, s, lineno) {
 
     // Convert to long
-    if (val > Sk.builtin.int_.threshold$ &&
+    if (val > Number.MAX_SAFE_INTEGER &&
```
Yeah. We do. And the function Sk.longFromStr handles how we deal with the compiling between py2 and py3 modes. 

If in py3 longFromString return an int vs py2 where it returns a long. 

Long is now in int.js at the bottom somewhere as a trivial subclass.

s-cork commented on src/abstract.js:
```diff
@@ -1039,6 +1256,7 @@ Sk.abstr.setUpInheritance = function (childName, child, parent) {
  * @param  {Object} self      The instance of the subclas.
  * @param  {...?} args Arguments to pass to the constructor.
  * @return {undefined}
+ * @deprecated
```
We don’t use it in the code base anymore. It was previously a way to implement the subclassing of builtins. 
That architecture is now part of each type object and dealt with in tp$new. 

When tp$new is called with a subtype then we decide there how to deal with it.

You can also see the generic handling in Sk.generic.new which is probably the equivalent of this function now.

s-cork commented on src/abstract.js:
```diff
@@ -1,181 +1,183 @@
     }
 };
-
+/**@suppress {checkTypes} */
```
It was because matmul is never defined in the codebase. 🤷‍♂️

s-cork replied:
resolved with ae9908c3551ae5645c3e67470b80d7a45fb70776
I hadn't defined these properly in `slotdefs.js` when i included these slots this was fixed. 

I have one suppression in `abstract.js` at present. I've added a comment about why that `checkType` seems to fail. 
I believe it's to do with order of imports but i'm not sure...

s-cork replied:
I think this is fully resolved now. I moved setUpBaseInheritance off of abstract and into object.js
It's one time function that didn't need to be in the `Sk.abstr` namespace. 
So now there are no `@suppress {checkTypes}` in `abstract.js`

s-cork commented on doc/ReferenceManual.rst:
```diff
@@ -161,7 +161,7 @@ Misc
     Slot("__init__", "tp$init", "init"),
     Slot("__str__", "tp$print", "print"),
-    Slot("__repr__", "tp$repr", "repr",
+    Slot("__repr__", "$r", "repr",
```
Changes here are a reflection of what exists. $r  for example is our equivalent of cpythons tp$repr. So I update this doc to reflect the code base as it currently stands. 

I replaced nb$nonzero with nb$bool since that’s a python2ism but we still support __nonzero__ as a dunder method in py2 mode it will just get the slot function nb$bool. And this switch is handled in the switch to py2 env function. 


I’ve tried to move more towards cpython hopefully since we now don’t have mp$del_subscript for example. Which was skulpt only. 


Curiously Cpython doesn’t have reflected slots like skulpt does. But the way cpython handles reflected number slots wouldn’t work the same way in skulpt. 

I’ve tried to document all the slots we support in the jsdocs. Or as comments in slotdefs.js.

s-cork replied:
Additional things. 

nb$divide is depreciated in cpython3
Because py2 has both nb$divide and nb$true_divide. 

I figured nb$divide is nicer and we can use that since we were already using it over nb$true_divide. But happy to move this in-line with cpython3. Its an easy change I think. 

nb$lng has a few versions in the code base but since nb$int_ matches its constructor name I figured nb$long should too

sq$ass_slice is not in cpython3 so I removed this from the codebase. 

I also use tp$as_number which is a cpython slot.  

Cpython has two slots tp$as_sequence and tp_as_mapping. 

I combine these as tp$as_sequence_or_mapping since skulpt handles these synonymously.

s-cork replied:
To resolve this I have updated this rst with the current slots in the codebase. 
The document was not previously up to date with master.
The names should now hopefully be correct at least.

albertjan replied:
thank you!

s-cork commented:
Agree that it shouldn’t be considered as mergable until after any non breaking prs can be merged. 

Any feedback and reviews before hand are much appreciated so I can work on it and get it to the place it needs to be. 

I really hope it doesn’t break anything. I’ve avoided touching the external apis like misceval.buildclass and builtin.func.

s-cork commented on src/constants.js:
```diff
@@ -1,52 +1,25 @@
 
 /* Constants used for kwargs */
 
-// Sk.builtin.int_
```
Yeah since doing int(“1”) is now handled completely differently. 

Before it would be dealt with by the applyOrSuspend architecture. Now it is handled by the cpython style architecture 
tp$call -> tp$new -> tp$init

at each stage the args, kwargs are passed and dealt with (mutable types typically handle args and kwargs in tp$init and immutable in tp$new). Every type object now gets type.prototype.tp$call method. 
It should make everything a lot cleaner.

s-cork commented on breakingchanges.md:
```diff
@@ -0,0 +1,153 @@
+**Additions**
+- `dict`, `tuple` are suspendable
+- `classmethod`, `property`, `staticmethod` have native skulpt implementations
+- `super` can now be unbound
```
See this previously notImplementedError 

https://github.com/skulpt/skulpt/blob/master/src/typeobject.js#L56

And this:

https://stackoverflow.com/questions/30190185/how-can-i-use-super-with-one-argument-in-python/30190341#30190341

I commented some tests for this in test_super. Though the tests needed changing since they  were out of date for py3. I tested the tests in python3 to make sure they were accurate. I should probably comment the py2 tests back out and find the appropriate py3 tests and move these to unit3/

s-cork replied:
to resolve this I added the stackoverflow link in the breaking changes. 
i updated 
`test_supers` in `unit/test_descr.py` with the up to date 2.7 tests
(this highlighted a small bug I had and fixed this)
I also included the cpython `test_super.py` file and uncommented all the tests that worked. 
commented tests were those that we don't yet support and one that I will raise an issue which is a closure issue.

s-cork commented on src/lib/datetime.py:
```diff
@@ -423,7 +421,7 @@ def _normalize_date(year, month, day, ignore_overflow=False):
 
 def _accum(tag, sofar, num, factor, leftover):
-    if isinstance(num, (int, long)):
+    if isinstance(num, (int)):
```
also worth noting that the behaviour remains the same and `long` is accounted for here since it's a subclass of `int` now

s-cork commented on src/builtin.js:
```diff
@@ -631,39 +432,37 @@ Sk.builtin.unichr = function unichr (x) {
 };
 
-Sk.builtin.hex = function hex (x) {
-    Sk.builtin.pyCheckArgsLen("hex", arguments.length, 1, 1);
```
resolving - the checkArgs can all be found in `builtindict.js` where the call signatures are now defined

s-cork commented on src/dict.js:
```diff
@@ -1,828 +1,710 @@
+        if (L === undefined) {
+            L = [];
+        }
+        Sk.asserts.assert(Array.isArray(L) && this instanceof Sk.builtin.dict, "bad call to dict constructor");
```
see commit 6a8804e9d293a8870060efc98091c53509496f40

s-cork commented on src/builtin.js:
```diff
@@ -984,16 +645,16 @@ Sk.builtin.jseval = function jseval (evalcode) {
 };
 
-Sk.builtin.jsmillis = function jsmillis () {
+Sk.builtin.jsmillis = function jsmillis() {
```
should i remove it from `builtindict`?

s-cork commented on src/builtin.js:
```diff
@@ -824,77 +511,48 @@ Sk.builtin.open = function open (filename, mode, bufsize) {
-    Sk.builtin.pyCheckArgsLen("isinstance", arguments.length, 2, 2);
-    if (!Sk.builtin.checkClass(type) && !(type instanceof Sk.builtin.tuple)) {
-        throw new Sk.builtin.TypeError("isinstance() arg 2 must be a class, type, or tuple of classes and types");
+const issubclass_multiple_inheritance = function (klass, base) {
```
turns out this function is a hangover from an earlier attempt at hacking inheritance. It's now no longer used. 
So i've removed it.

s-cork commented:
Updates on this branch - I have merged in most of the latest changes and just need to merge in the last handful of prs before I make a push to this pr.

The latest version of the branch can be found here: https://github.com/s-cork/skulpt/tree/bigpr_current_state


long integers now use javascript's native `bigint` if available (which is now supported in the latest versions of all [major browsers](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/BigInt#Browser_compatibility)!) - for older browsers we use the [`JSBI`](https://github.com/GoogleChromeLabs/jsbi) library. 
To make this work I take advantage of the JSBI api and rewrite the functional approach to do the obvious thing e.g. `JSBI.add = (x,y) => x + y;`
exact implementation can be found here: https://github.com/s-cork/skulpt/blob/bigpr_current_state/support/polyfills/JSBI.js


I've started to include as many cpython tests as I can in order to make sure I am:
1. not introducing new bugs that were not covered by the test suite
2. fixing existing bugs we didn't know were there. 

I have cpython full(ish) coverage of: `tuple`, `list`, `mappingproxy`, `dict`, `set`, `frozenset`,  `descriptors` - including tests that subclass these builtins.
yet to include full coverage of other type objects. 
I have now touched the codebase for every type object and converted to the proposed api (except `file`)

I've introduced support for the union operator on dicts: as per [PEP 584](https://www.python.org/dev/peps/pep-0584/ )
the `|` operator is now also valid for `dict.keys()` and `dict.items()` which will return `sets` as per cpython. 

I think a version of this pr is already in production in @acbart's [blockpy](https://github.com/blockpy-edu/blockpy )

performance wise this branch is around 2-3x faster than master for `field.py` 
(#1143 - cut master times in half and there have been a couple of performance improvements from recent commits on master)

I adjust the dict entries from `{lhs: pyobject, rhs: pyobject}` to be an array. 
The reason for this idea is that we now have fast internal iteration paths for `dict` objects
```javascript
// fast path for dict iteration
dict.$items().forEach(([key, val]) => {/* do something with the dict items */})
```
I take advantage of this fast path if a few select places in the codebase e.g. `Counter` which now overrides the `number` slots.

I also introduce a `quick$lookup` for looking up a `str` key in a python dictionary. 
```javascript
Sk.builtin.dict.prototype.quick$lookup = function (pyName) {
    var key_hash = pyName.$savedKeyHash;
    var item = this.entries[key_hash];
    if (item !== undefined) {
        return item[1];
    }
    return;
};
```
this fastpath is used in the obvious places in the codebase. 

I've also rewritten `operator` to use the `BuiltinFunctionOrMethodType` and included `itemgetter`, `attrgettr` and `methodcaller` as native skulpt class objects. 

all `itertools` objects are now `typeobjects` rather than hacked generators.

s-cork commented:
I've merged up to master. I'll leave this branch until anyone is brave enough to review. 😅

alexwenbj commented:
`Sk.misceval.callsim(Sk.builtin.float_, 2.1)` and `Sk.misceval.callsim(Sk.builtin.float_, "2.1")` both raised ` float() argument must be a string or a number` in this PR.

s-cork commented:
I can add that to breaking changes. 
But that's somewhat intentional.

Since `Sk.buitin.float` has a `__new__/tp$new` method the use of `Sk.misceval.callsim` will call the `Sk.builtin.float`'s `__new__` implementation (rather than the `Sk.builtin.float` constructor as before) 
and so (when using `Sk.misceval.callsim`) it is expecting to be called via python rather than natively in javascript. 

to call the float constructor in javascript this PR would then require ether
- `new Sk.builtin.float(2.1)` or 
- `Sk.ffi.remapToPy(2.1)`

alexwenbj commented:
@s-cork I am using the third party numpy lib, here is the code: https://github.com/ebertmi/skulpt_numpy/blob/master/numpy/__init__.js#L2729
cause the data type is uncertain, so it uses `callsim()` function.

alexwenbj commented:
And yes, we can replace the `callsim() `invocation with `Sk.ffi.remapToPy() `here, but maybe there are other files using `callsim()`,so if it is possible, it is better make `callsim()` works in the old way, I think.
Thank you.

meredydd commented:
Despite my usual advocacy of compatibility I disagree.

The `call()` functions make a Python call using Python objects. It was a source of both bugs and performance problems that _some_ unpredictable subset of those Python functions could also take Javascript arguments.

As of the last release, we're officially in a window for breaking changes, and this is a pretty easy change to search for in third-party code (search for every use of the call functions, make sure they're taking Python-valued args, and if you're instantiating Skulpt primitives, use `new`).

alexwenbj commented:
I agree @meredydd , maybe we should list all the breaking changes to inform the developers how to modify their own codes or the third party libs.
By the way, this PR is a great improvement, and thank @s-cork very much.

bennorth commented:
I can add another success report for this PR.

A colleague and I are working [an educational Python environment called **Pytch**](https://pytch.org/), which aims to be a stepping stone between [MIT's Scratch](https://scratch.mit.edu/) and Python.  We've built on Skulpt, making heavy use of its suspensions to implement event-triggered concurrency.

If you haven't come across Scratch, it's an excellent system for learning programming, where you write programs by dragging/dropping blocks.  There's a nice runtime with interactive sprites, built-in graphics, sounds, etc., so it's easy for novice programmers to concentrate on the interesting bits.  Once they're skilled in Scratch, people often want to move on to Python.  But then not only do they have to deal with a whole new way of expressing their programs (as text, getting all the syntax right), but they also have to deal with a whole new runtime, leaving behind their sprites, sounds, and so on.  Pytch lets them stay in that world, keeping all the understanding they've built up about it, and concentrate just on the change from block-based to text-based programming.

I made an experimental branch of Pytch, and merged in @s-cork's branch. I had to make a handful of changes to Pytch: using `new` for builtins was mechanical; I'd made a few assumptions about Skulpt internals which needed updating; I changed to the new `buildNativeClass()` in some cases. Then Pytch's own suite of unit tests all passed, as did its browser-based UI tests.

More info:
* [Pytch](https://pytch.org/) — to see how we're using Skulpt, [choose a tutorial](https://pytch.org/app/tutorials/) and click the "Try the finished project!" button.
* [Pytch VM source](https://github.com/pytchlang/pytch-vm) — this is a fork of the Skulpt codebase, and adds logic for event-driven concurrency and other Scratch-like concepts. The handful of changes I needed to make post-merge of this PR are in the branch [expt/try-s-cork-descriptors-rewrite](https://github.com/pytchlang/pytch-vm/commits/expt/try-s-cork-descriptors-rewrite).

meredydd reviewed (approved):
Reviewed 1:1 with @s-cork, tested, and in production on Anvil for months now!

meredydd commented:
Hi all,

I think it's time to get this sucker merged! I have gone over it with a fairly fine tooth-comb with @s-cork, and all my feedback is already incorporated in the current state of the PR.

We have been running it in production at Anvil for months, and it's getting traction elsewhere, too: At least @acbart's BlockPy, and @bennorth and @alexwenbj's projects are using it in production.

Waiting for it to land is also blocking up Skulpt development - either PRs are stacking on top of it (#1240, #1241) or others are nervous to contribute to something that's about to be refactored.

It's a big win for python compatibility, and the performance gain is simply eye-popping (@rixner should be happy ;) ). It translates beyond benchmarks, too - since we shipped it, we got a lot of "Is it just me, or has everything got noticeably faster?" on the Anvil forums.

@bnmnetp I think we should just hit the button, and as soon as possible.

LeszekSwirski commented:
FWIW I've looked at the getattr and call internals (mostly with an eye on performance gotchas more than correctness) and those looked good to me.

bnmnetp commented:
Trying this out on my Fork.  Things looked very promising until I get to document.

This simple program:
```
import document

t = document.getElementById('text1')
print('value = ', t.value)
```

Which has always worked fine now fails with `TypeError: could not convert object of type '' to str on line 3`

```
    mod.getElementById = new Sk.builtin.func(function (id) {
        var result = document.getElementById(id.v);
        if (result) {
            return Sk.misceval.callsimArray(mod.Element, [result]);
        }
        return Sk.builtin.none.none$;
    });
```

What needs to be updated in the definition to get it to work again?
​

bnmnetp commented:
Ok, so I've tracked this down a bit further.  The error is coming from a line:

```
new Sk.builtin.str(self.checked)
```
where self.checked evaluates to `false`

bnmnetp commented:
Nevermind -- It turns out that I should have been making a new bool not a new string.  In my defense this particular line of code is at least 8 years old.  But oddly has been working many thousands of times a day until I test merged this PR. 😄

bnmnetp commented:
And to be fair `new Sk.builtin.str(Sk.builtin.bool(false))` works just like it should.

Merging...
